### PR TITLE
Refactoring report_method_error 

### PR DESCRIPTION
**Pull request by chenyukang**
While working on #105732, I found it's hard to follow this long function,
so I tried to make it shorter.

It's not easy for code reviewing, since so many lines of code changes, 
but only the positions are changed.

Generally, I extract two sub-methods from `report_method_error`:

https://github.com/rust-lang/rust/blob/397b66e77b279de5006facf87979f9ecff5c7f87/compiler/rustc_hir_typeck/src/method/suggest.rs#L117
to `note_candidates_on_method_error`

And this long block:
https://github.com/rust-lang/rust/blob/397b66e77b279de5006facf87979f9ecff5c7f87/compiler/rustc_hir_typeck/src/method/suggest.rs#L263

to `report_no_match_method_error`.

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
This could use a rebase. I'll approve after that.

**Comment by chenyukang** [comment]
emmm, conflict again...
@rustbot ready
